docs/design/design.md

```diff
@@ -0,0 +1,131 @@
+- Well-documented,  we organize the API with swagger and keep the API synchronized with the documentation through swagger generation tools.
+
+### Reference Docs
+
+For API definition, please refer to [swagger.yml](https://github.com/dragonflyoss/Dragonfly/blob/master/apis/swagger.yml)
+
+For API document, please refer to [apis.md](https://github.com/dragonflyoss/Dragonfly/blob/master/docs/apis.md)
+
+## Architecture
+
+In DragonFly, we have two roles:
+
+- Supernode, a SuperNode is a long-time process that plays the following roles:
+  - A tracker and scheduler in the P2P network that chooses the appropriate downloading net-path for each peer.
+  - A CDN server that caches downloaded data from the source to avoid downloading same files repeatedly.
+- Dfclient, Dfclient contains two components:
+  - Dfget, Dfget is the Dragonfly client used for downloading files. It's similar to wget. Meanwhile, it also plays the role of a peer with uploader, which can transfer data between each other in the P2P network.
+  - Dfdaemon, Dfdaemon is only used for pulling images. It establishes a proxy between dockerd/pouchd and registry.Dfdaemon filters out layer fetching requests from all requests sent by dockerd/pouchd when pulling images, then uses Dfget to download these layers.
+
+As mentioned above, we use the following picture to make it easier for you to understand.
+
+![architecture.png](../images/design/architecture.png)
+
+## Workflows
+
+### Dfget and Supernode's communication workflows
+
+When Dfget try to download files/images from Supernode, there are three steps to follow:
+
+1. Register to a Supernode
+2. Download from the p2p network in pieces with a loop
+3. ServiceDown
+
+Below is a diagram illustrating that how Dfget download a file from Supernode.
+
+![dfclient_supernode.png](../images/design/dfget_supernode.png)
+
+#### Register to a Supernode
+
+1. Dfget should launch a server as a peer and refister to Supernode.
+2. Dfget should tell the Supernode some info about files/images to be downloaded.
+
+#### Download from the p2p network in pieces with a loop
+
+The Dfget download is not a one-time work, it will download a piece at a time with a loop until the download is complete. It is mainly divided into the following three steps:
+
+- Get peers information with the pieces to be downloaded.
+- Download the pieces from the specified peers.
+- Update the pieceInfo to tell Supernode that it has downloaded the piece for two things:
+  - Provides a download service for other nodes in the network that need the piece
+  - Get the information about other pieces that need to be downloaded
+
+#### ServiceDown
+
+When Dfget lasts in five minutes without downloading or uploading tasks, uploader of Dfget will automatically sends a serviceDown request to Supernode.
```

Comment:
I think this may actually be a problem if we are not careful enough: Is supernode capable to deregister this peer automatically. (It is not a good idea to rely on clients for this process.)